### PR TITLE
[fix] don't expose unique_dim in torch

### LINKED ISSUE #62793: AttributeError: module 'torch' has no attribute 'unique_dim'
## 🐛 Bug

When using `from torch import *`, the error `AttributeError: module 'torch' has no attribute 'unique_dim'` appears.

## To Reproduce

1. install pytorch 1.9.0
2. from torch import *

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Pro
GCC version: (MinGW.org GCC-6.3.0-1) 6.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.7.6 (default, Jan  8 2020, 20:23:39) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19041-SP0
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: GPU 0: GeForce GTX 1060 6GB
Nvidia driver version: 461.40
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.1\bin\cudnn_ops_train64_8.dll
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.1
[pip3] numpydoc==1.1.0
[pip3] pytorch-lightning==0.5.2.1
[pip3] torch==1.9.0+cu111
[pip3] torchaudio==0.9.0
[pip3] torchvision==0.10.0+cu111
[conda] blas                      1.0                         mkl    defaults
[conda] cudatoolkit               11.1.1               heb2d755_7    conda-forge
[conda] mkl                       2020.2                      256    defaults
[conda] mkl-service               2.3.0            py37h196d8e1_0    defaults
[conda] mkl_fft                   1.3.0            py37h46781fe_0    defaults
[conda] mkl_random                1.1.1            py37h47e9c7a_0    defaults
[conda] numpy                     1.21.1                   pypi_0    pypi
[conda] numpy-base                1.18.5           py37hc3f5095_0    defaults
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1    defaults
[conda] pytorch                   1.9.0           py3.7_cuda11.1_cudnn8_0    pytorch
[conda] pytorch-lightning         0.5.2.1                  pypi_0    pypi
[conda] torch                     1.2.0                    pypi_0    pypi
[conda] torchaudio                0.9.0                    pypi_0    pypi
[conda] torchvision               0.10.0+cu111             pypi_0    pypi

## Additional context

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #62793

This is mostly a quick fix. I think the more correct fix could be updating `unique_dim` to `_unique_dim` which could be BC-breaking for C++ users (🤔 maybe). Maybe something else I am missing.

~~Not sure how to add a test for it.~~ Have tested it locally.

We can add a test like following. Tested this locally, it fails currently but passes with the fix.
```python
        def test_wildcard_import(self):
            exec('from torch import *')

```

### CODE REVIEW

**Comment by zou3519** [comment]
cc @zasdfgbnm @vitaly-fedyunin you've done a lot of work on unique_dim in the past (see stack at https://github.com/pytorch/pytorch/pull/18649), what do you think about this change?

**Review by zasdfgbnm — COMMENTED** [review]
Would it help if we add `unique_dim` to this list?https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_python_functions.py#L72

**Comment by kshitij12345** [comment]
> Would it help if we add unique_dim to this list?

AFAIK, we won't get Python bindings at all with this. So we can't use `_VF.unique_dim` in functional.py. (Will try it though).

**Comment by kshitij12345** [comment]
With this
```patch
diff --git a/tools/autograd/gen_python_functions.py b/tools/autograd/gen_python_functions.py
index d1fb70c8ab..14b91f5718 100644
--- a/tools/autograd/gen_python_functions.py
+++ b/tools/autograd/gen_python_functions.py
@@ -93,7 +93,7 @@ _SKIP_PYTHON_BINDINGS = [
     'data', 'is_leaf', 'output_nr', '_version', 'requires_grad_', 'retains_grad', 'set_',
     '_fw_primal', 'fake_quantize_per_tensor_affine_cachemask',
     'fake_quantize_per_channel_affine_cachemask',
-    '_reshape_alias',
+    '_reshape_alias', 'unique_dim'
 ]
```

There are other errors on import ops,

<details>

```
>>> import torch
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/__init__.py", line 678, in <module>
    import torch.nn.quantizable
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/nn/quantizable/__init__.py", line 1, in <module>
    from .modules import *  # noqa: F403
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/nn/quantizable/modules/__init__.py", line 1, in <module>
    from .activation import MultiheadAttention
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/nn/quantizable/modules/activation.py", line 4, in <module>
    import torch.nn.quantized as nnq
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/nn/quantized/__init__.py", line 1, in <module>
    from .modules import *  # noqa: F403
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/nn/quantized/modules/__init__.py", line 4, in <module>
    from .activation import ReLU6, Hardswish, ELU, LeakyReLU, Sigmoid
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/nn/quantized/modules/activation.py", line 2, in <module>
    import torch.nn.quantized.functional
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/nn/quantized/functional.py", line 9, in <module>
    from torch.jit.annotations import BroadcastingList2
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/jit/__init__.py", line 21, in <module>
    from torch.jit._script import (
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/jit/_script.py", line 22, in <module>
    from torch.jit._recursive import ScriptMethodStub, wrap_cpp_module, infer_methods_to_compile, _compile_and_register_class
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/jit/_recursive.py", line 13, in <module>
    from torch.jit._builtins import _find_builtin
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/jit/_builtins.py", line 104, in <module>
    (torch._VF.unique_dim, "aten::unique_dim"),
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/site-packages/torch/_VF.py", line 26, in __getattr__
    return getattr(self.vf, attr)
AttributeError: 'torch._C._VariableFunctionsClass' object has no attribute 'unique_dim'
```

</details>

**Review by zasdfgbnm — APPROVED** [review]
Thanks for fixing!

File: torch/__init__.py [hunk 1]
```diff
@@ -609,8 +609,14 @@ def manager_path():
     # PR #43339 for details.
     from torch._C._VariableFunctions import *  # type: ignore[misc] # noqa: F403
 
+# Ops not to be exposed in `torch` namespace,
+# mostly helper ops.
+PRIVATE_OPS = (
+    'unique_dim',
+)
+
```
Comment (zou3519) [inline torch/__init__.py:617]: I'm trying to understand the issue here. We don't want to expose torch.unique_dim. However, we want torch._C._VariableFunctions.unique_dim to work. Therefore we need to enable Python bindings for "unique_dim" and delete it here.

The reason why we can't rename "unique_dim" to "_unique_dim" is for BC reasons. Does that sound right?

**Comment by zou3519** [comment]
@kshitij12345 I think we should pursue renaming unique_dim to _unique_dim.

`torch.unique` is already pretty messed up over in C++-land (in C++ right now a user either calls torch::_unique2 or torch::unique_dim). If we want to preserve BC for the C++ API we can just add a torch::unique_dim as a function in C++ (maybe somewhere near https://github.com/pytorch/pytorch/blob/de7ae9e9b6a8537ca06727f76363002350a7bacc/torch/csrc/api/include/torch/types.h#L31).

File: torch/__init__.py (hunk 1 above)
Comment (kshitij12345) [reply torch/__init__.py:617]: Yes. That's about it

**Comment by zou3519** [comment]
> @kshitij12345 I think we should pursue renaming unique_dim to _unique_dim.

So if you agree (feel free to disagree!) we should open a new PR performing the rename and leave this one open just in case we figure out that the BC concerns are too much to handle at the moment

**Comment by kshitij12345** [comment]
On it. 
However I was wondering if we should use this PR to rename ops like `_unique_dim`, `_unique2`, `_cat`, etc from torch namespace?

**Review by zou3519 — APPROVED** [review]

**Comment by zou3519** [comment]
@kshitij12345 could you rebase this please when you get a chance? Thank you

**Comment by kshitij12345** [comment]
@zou3519 Done :)

**Pull request closed without merging**